.pre-commit-config.yaml

```diff
@@ -1,9 +1,10 @@
 repos:
-  - repo: local
+  - repo: "local"
     hooks:
-    - id: codechecks
-      name: codechecks
+    - id: "codechecks"
+      name: "codechecks"
       stages: ["push"]
+      language: "system"
+      files: ".py$"
       entry: "tox -e codechecks"
-      language: system
-      pass_filenames: false
+      pass_filenames: true
```

Comment:
Perhaps I am confused about pre-commit... I am not sure if pre-commit has the intended effect.

I am not sure if it is a configuration problem, partly because `pre-commit-validate-config .pre-commit-config.yaml` isn't suggesting anything.

I did the below things, with some debug prints added in our checker scripts:

```
$ pre-commit --version
pre-commit 1.21.0
$ pre-commit install
pre-commit installed at .git/hooks/pre-commit
$ pre-commit run 
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/sajith/.cache/pre-commit/patch1606750288.
[INFO] Restored changes from /home/sajith/.cache/pre-commit/patch1606750288.
$  git commit -m "Add debug prints"
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/sajith/.cache/pre-commit/patch1606750320.
[INFO] Restored changes from /home/sajith/.cache/pre-commit/patch1606750320.
On branch 3515.partial-pre-commit
Your branch is up to date with 'origin/3515.partial-pre-commit'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   misc/coding_tools/check-debugging.py
	modified:   misc/coding_tools/check-umids.py
	modified:   misc/coding_tools/find-trailing-spaces.py
 git commit -m "Add debug prints"
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/sajith/.cache/pre-commit/patch1606750320.
[INFO] Restored changes from /home/sajith/.cache/pre-commit/patch1606750320.
On branch 3515.partial-pre-commit
Your branch is up to date with 'origin/3515.partial-pre-commit'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   misc/coding_tools/check-debugging.py
	modified:   misc/coding_tools/check-umids.py
	modified:   misc/coding_tools/find-trailing-spaces.py
$ git commit -a -m "Add debug prints"
[3515.partial-pre-commit 0488c10d7] Add debug prints
 3 files changed, 5 insertions(+)
$ git push mine 3515.partial-pre-commit 
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Delta compression using up to 4 threads
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 721 bytes | 360.00 KiB/s, done.
Total 7 (delta 6), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (6/6), completed with 6 local objects.
remote: 
remote: Create a pull request for '3515.partial-pre-commit' on GitHub by visiting:
remote:      https://github.com/sajith/tahoe-lafs/pull/new/3515.partial-pre-commit
remote: 
To github.com:sajith/tahoe-lafs.git
 * [new branch]          3515.partial-pre-commit -> 3515.partial-pre-commit
```

It doesn't seem that pre-commit ran `tox -e codechecks` on `git push`.  

It is odd that pre-commit did run at `git commit -m`, but it did not run at `git commit -a -m`.   

The fact a `.git/hooks/pre-commit` was installed, but not `.git/hooks/pre-push` must be a pre-commit bug?  Or is it just the way it is supposed to be?

Reply:
You can choose which kind of hook to install it as.  I did this because I don't want to wait for it to run for every single commit:

```
pre-commit install --hook-type pre-push
```

I think pre-commit has a bunch of usability issues (though maybe these are forced on it by git, I dunno).  But we're just trying to benefit from it, I'm not interested in trying to fix it...

Maybe our own docs about how we expect Tahoe-LAFS devs to use this could be clearer, though?

Reply:
Aaah okay.  I thought that the `stages: ["push"]` line will do something to that effect. :-)

I think documentation will help.  Per docs/developer-guide, running `make install-vcs-hooks` is the way to install git hooks, but that requires py36 tox environment, which I guess would be rather rare to come by.